File: deploy/cdk/marble-web-kiosk-export-pipeline/lib/build-project.ts
import cdk = require('@aws-cdk/core')
import codebuild = require('@aws-cdk/aws-codebuild')
import { BuildEnvironmentVariableType } from '@aws-cdk/aws-codebuild'
import { Role } from '@aws-cdk/aws-iam'

export enum BuildProjectType {
  BUILD,
  DEPLOY,
}

export interface BuildProjectProps {
  projectType: BuildProjectType,
  role: Role,
  stage: String,
  context?: { [key: string]: any },
}

const getContextCliArguments = (props: BuildProjectProps) => {
  let output = ''
  // if context is undefined, create an empty object so it doesn't crash
  const context: { [key: string]: any } = props.context || {}
  for (const key in context) {
    if (key) {
      output += `-c ${key}="${context[key]}"`
    }
  }
  return output
}

const getPhases = (props: BuildProjectProps) => {
  switch (props.projectType) {
    case BuildProjectType.BUILD:
      return {
        install: {
          'runtime-versions': {
            nodejs: 10,
            python: 3.7,
          },
          commands: [
            'echo "Ensure that the codebuild directory is executable"',
            'chmod -R 755 ./scripts/codebuild/*',
            './scripts/codebuild/install.sh',
            'chmod -R 755 ./local_install.sh',
            './local_install.sh',
          ],
        },
        pre_build: {
          commands: [`./scripts/codebuild/pre_build.sh`],
        },
        build: {
          commands: [`./scripts/codebuild/build.sh ${getContextCliArguments(props)}`],
        },
        post_build: {
          commands: ['./scripts/codebuild/post_build.sh'],
        },
      }
    case BuildProjectType.DEPLOY:
      return {
        install: {
          'runtime-versions': {
            nodejs: 10,
            python: 3.7,
          },
          commands: [
            'echo "Ensure that the codebuild directory is executable"',
            'chmod -R 755 ./scripts/codebuild/*',
            './scripts/codebuild/install.sh',
          ],
        },
        build: {
          commands: [`./scripts/codebuild/deploy.sh ${getContextCliArguments(props)} --verbose`],
        },
      }
  }
}

export const BuildProject = (scope: cdk.Construct, props: BuildProjectProps) => {
  return new codebuild.PipelineProject(scope, `${props.stage}${Object.values(BuildProjectType)[props.projectType]}Project`, {
    environment: {
      buildImage: codebuild.LinuxBuildImage.STANDARD_2_0,
    },
    environmentVariables: {
      'CI': { value: 'true', type: BuildEnvironmentVariableType.PLAINTEXT },
      'STAGE': { value: props.stage, type: BuildEnvironmentVariableType.PLAINTEXT },
    },
    role: props.role,
    buildSpec: codebuild.BuildSpec.fromObject({
      version: '0.2',
      phases: getPhases(props),
      artifacts: {
        'files': [
          'dist/**/*',
          'scripts/codebuild/**/*'
        ],
      },
    }),
  })
}

File: deploy/cdk/marble-elasticsearch-pipeline/marble_elasticsearch_pipeline/build_project.py
from aws_cdk import core
import aws_cdk.aws_codebuild as codebuild


class BuildProject():
    def __init__(self, scope: core.Construct, role, stage, context=None):
        self.scope = scope
        self.role = role
        self.stage = stage
        self.context = context

    def pipeline_project(self):
        project_name = f"{self.stage}Project"
        env = {'buildImage': codebuild.LinuxBuildImage.STANDARD_2_0}
        env_vars = {
            'CI': {'value': 'true', type: codebuild.BuildEnvironmentVariableType.PLAINTEXT},
            'STAGE': {'value': self.stage, type: codebuild.BuildEnvironmentVariableType.PLAINTEXT},
        }
        artifacts = {
            'files': [
                'cdk.out/*',
                'scripts/codebuild/**/*'
            ],
        }
        return codebuild.PipelineProject(self.scope, project_name, role=self.role, environment=env,
                    build_spec=codebuild.BuildSpec.from_object({'version': '0.2', 'phases': self._get_phases(), 'artifacts': artifacts}))

    def _get_phases(self):
        return {
            'install': {
                'runtime-versions': {
                    'python': 3.7,
                },
                'commands': [
                    'echo "Ensure that the codebuild directory is executable"',
                    'chmod -R 755 ./scripts/codebuild/*',
                    './scripts/codebuild/install.sh'
                ],
            },
            'pre_build': {
                'commands': [f"./scripts/codebuild/pre_build.sh {self.context['es_stackname']}-{self.stage}"],
            },
            'build': {
                'commands': ['./scripts/codebuild/deploy.sh'],
            },
            'post_build': {
                'commands': ['./scripts/codebuild/post_build.sh'],
            },
        }
